rs/modelator/src/model/language/tla/json/parser.rs

```diff
@@ -1,65 +1,126 @@
 
 pub(crate) fn parse_state(i: &str) -> IResult<&str, JsonValue> {
     map(
         preceded(
             opt(delimited(multispace0, complete(tag("/\\")), multispace0)),
             separated_list0(
                 delimited(multispace0, complete(tag("/\\")), multispace0),
-                parse_var,
+                parse_assignment,
             ),
         ),
-        |value| JsonValue::Object(value.into_iter().collect()),
+        |pairs| {
+            let j = JsonValue::Object(pairs.into_iter().collect());
+            println!("{}", serde_json::to_string_pretty(&j).unwrap());
+            j
+        },
     )(i)
 }
 
-fn parse_var(i: &str) -> IResult<&str, (String, JsonValue)> {
+fn parse_assignment(i: &str) -> IResult<&str, (String, JsonValue)> {
     delimited(
         multispace0,
         separated_pair(
-            parse_identifier,
+            parse_identifier_as_string,
             delimited(multispace0, char('='), multispace0),
-            parse_any_value,
+            parse_anything,
         ),
         multispace0,
     )(i)
 }
 
 // TODO: what else can TLA var idenfitiers have?
-fn parse_identifier(i: &str) -> IResult<&str, String> {
+fn parse_identifier_as_string(i: &str) -> IResult<&str, String> {
     map(
         many1(satisfy(|c| c.is_alphanumeric() || "_-".contains(c))),
         |value| value.into_iter().collect(),
     )(i)
 }
 
-fn parse_identifiers_as_values(i: &str) -> IResult<&str, JsonValue> {
-    map(parse_identifier, JsonValue::String)(i)
+fn parse_identifier(i: &str) -> IResult<&str, JsonValue> {
+    map(parse_identifier_as_string, |x| json!(x))(i)
+}
+
+fn parse_any_value_consuming(i: &str) -> IResult<&str, JsonValue> {
+    alt((
+        parse_bool,
+        parse_number,
+        parse_string,
+        parse_identifier,
+        parse_set,
+        parse_sequence,
+        parse_record,
+    ))(i)
+}
+
+// TODO: add more TLA+ operators? support prefix operators
+fn parse_operator(i: &str) -> IResult<&str, JsonValue> {
+    map(alt((tag("\\in"), tag("\\notin"))), |x| {
+        json!({ "operator": x })
+    })(i)
+}
+
+fn parse_expression(i: &str) -> IResult<&str, JsonValue> {
+    map(
+        delimited(
+            multispace0,
+            pair(
+                alt((
+                    delimited(char('('), parse_anything, char(')')),
+                    parse_any_value_consuming,
+                )),
+                many1(pair(
+                    delimited(multispace1, parse_operator, multispace1),
+                    alt((
+                        delimited(char('('), parse_anything, char(')')),
+                        parse_any_value_consuming,
+                    )),
+                )),
+            ),
+            multispace0,
+        ),
+        |(f, r)| {
+            let mut a = vec![f];
+            r.into_iter().for_each(|(p, q)| {
+                a.push(p);
+                a.push(q);
+            });
+            json!({ "expression": a })
```

Comment:
I added these tags like `expression`, `functions`, `record`, `string`, `set`, `sequence` to mark objects with types. Does this make sense?